lib/parse.js

```diff
@@ -57,8 +57,19 @@ var parseValues = function parseQueryStringValues(str, options) {
 
     var cleanStr = options.ignoreQueryPrefix ? str.replace(/^\?/, '') : str;
     cleanStr = cleanStr.replace(/%5B/gi, '[').replace(/%5D/gi, ']');
+    
+    // Calculate limit and split query string
     var limit = options.parameterLimit === Infinity ? undefined : options.parameterLimit;
-    var parts = cleanStr.split(options.delimiter, limit);
+    var parts = cleanStr.split(options.delimiter);
+    
+    // Check if parts exceed the parameter limit and handle it
+    if (parts.length > limit) {
+        if (options.throwOnParameterLimitExceeded) {
+            throw new Error(`Parameter limit exceeded. Only ${limit} parameters allowed.`);
+        }
+        parts = parts.slice(0, limit); // Keep only up to the limit
+    }
+    
```

Comment:
```suggestion
    var parts = cleanStr.split(
      options.delimiter,
      options.throwOnParameterLimitExceeded ? limit + 1 : limit
    );
    
    // Check if parts exceed the parameter limit and handle it
    if (options.throwOnParameterLimitExceeded && parts.length > limit) {
        throw new Error(`Parameter limit exceeded. Only ${limit} parameters allowed.`);
    }
    
```
this way there's no perf hit on the non-throw path, and it's faster on the throw path (because it only needs to check "limit + 1" (specifically, without the limit i think it would be a DOS vulnerability and require a CVE)